src/Run.hs

```diff
@@ -1,17 +1,57 @@
+{-# LANGUAGE OverloadedStrings #-}
 module Run where
 
 import           Control.Exception
+import           Control.Concurrent
+import           Control.Applicative
+import           Control.Monad (void, forever)
 import           Data.Foldable
 import           Data.List
+import           Data.Time.Clock
+import           System.FSNotify (withManager, watchTree)
 
 import           Interpreter (Interpreter)
 import qualified Interpreter
 
-run :: [String] -> String -> IO ()
-run args input = bracket (Interpreter.new args) Interpreter.close $ \interpreter -> do
-  trigger interpreter >> return ()
+data Event = Event UTCTime | Done
+
+emitEvent :: Chan Event -> IO ()
+emitEvent chan = (Event <$> getCurrentTime) >>= writeChan chan
+
+watchFiles :: Chan Event -> IO ()
+watchFiles chan = void . forkIO $ do
+  withManager $ \manager -> do
+    _ <- watchTree manager "." (const True) (const $ emitEvent chan)
+    forever $ threadDelay maxBound
+
+watchInput :: Chan Event -> String -> IO ()
+watchInput chan input = void . forkIO $ do
   forM_ (lines input) $ \_ -> do
-    trigger interpreter
+    emitEvent chan
+  writeChan chan Done
+
+run :: [String] -> String -> IO ()
+run args input = do
+  chan <- newChan
+  watchFiles chan
+  watchInput chan input
+  bracket (Interpreter.new args) Interpreter.close (processEvents chan)
+
+processEvents :: Chan Event -> Interpreter -> IO ()
+processEvents chan interpreter = do
+  t <- getCurrentTime
+  emitEvent chan
+  go t
+  where
+    go t0 = do
+      event <- readChan chan
+      case event of
+        Done -> return ()
+        Event t | t0 < t -> do
```

Comment:
Hmm, but I see that emtying the channel is not that straight-forward. `isEmptyChan` is deprecated.